### PR TITLE
Add ERB Support for SQL definitions

### PR DESCRIPTION
**Pull request by gmhawash**
Allow for ERB interpolation in SQL definition files.  This is helpful when SQL definition relies on some ruby constants or methods.

### CODE REVIEW

**Comment by teoljungberg** [comment]
I have never seen this in real life practice, can you cite a project or two that uses this?

**Comment by gmhawash** [comment]
Our application uses a private gem that exposes a set of transaction codes.  Every once in a while we update the gem (based on customer requirements) and the transaction codes are used in SQL functions, views and queries.  So for us, rather than listing the transactions codes all over the place; we do this inside these SQL entities:

```
SELECT ....
WHERE transactions.code  IN (<%= TransactionGem.payment_codes.join(',') %>) OR transactions.code in (<%= TransactionGem.jouranl_entry_codes.join(',') %>)
```

And when we update the transaction gem; we just rerun the migration to recreate the SQL functions and views and that updates them properly.

**Comment by teoljungberg** [comment]
I see! Thanks for context.

I'm going through a maintenance of the project to cut out old and stale versions of ruby and rails. And will cut a release for that before I come back to this.

File: lib/fx/version.rb [hunk 1]
```diff
@@ -1,4 +1,4 @@
 module Fx
   # @api private
-  VERSION = "0.7.0"
+  VERSION = "1.7.1"
```
Comment (teoljungberg) [inline lib/fx/version.rb:3]: You don't need to change this, I'll do that once a release is ready.

```suggestion
  VERSION = "0.7.0"
```

Or to be on top of master

```suggestion
  VERSION = "0.8.0"
```

File: spec/fx/definition_spec.rb [hunk 2]
```diff
@@ -29,6 +29,39 @@
...
+            RETURNS text AS $$
+            BEGIN
+                RETURN '<%= function %>';
+            END;
+            $$ LANGUAGE plpgsql;
+          <% end %>
+        EOS
+
+        erb_sql_definition = (<<-EOS).delete("\n").squeeze(' ').strip
+            CREATE OR REPLACE FUNCTION test()
```
Comment (teoljungberg) [inline spec/fx/definition_spec.rb:45]: Swap this to 2 space indentation

File: spec/fx/definition_spec.rb [hunk 3]
```diff
@@ -29,6 +29,39 @@
...
+            BEGIN
+                RETURN 'test';
+            END;
+            $$ LANGUAGE plpgsql;
+            CREATE OR REPLACE FUNCTION west()
+            RETURNS text AS $$
+            BEGIN
+                RETURN 'west';
+            END;
+            $$ LANGUAGE plpgsql;
```
Comment (teoljungberg) [inline spec/fx/definition_spec.rb:56]:
```suggestion
          CREATE OR REPLACE FUNCTION test()
          RETURNS text AS $$
          BEGIN
              RETURN 'test';
          END;
          $$ LANGUAGE plpgsql;
          CREATE OR REPLACE FUNCTION west()
          RETURNS text AS $$
          BEGIN
              RETURN 'west';
          END;
          $$ LANGUAGE plpgsql;
```

**Comment by gmhawash** [comment]
I pushed a commit to resolve these issues.

File: spec/fx/definition_spec.rb [hunk 4]
```diff
@@ -8,7 +8,7 @@
           CREATE OR REPLACE FUNCTION test()
           RETURNS text AS $$
           BEGIN
-              RETURN 'test';
```
Comment (teoljungberg) [inline spec/fx/definition_spec.rb:11]: Thanks for fixing these too 👏🏻

File: spec/fx/definition_spec.rb [hunk 5]
```diff
@@ -29,13 +29,46 @@
...
+          BEGIN
+            RETURN 'west';
+          END;
+          $$ LANGUAGE plpgsql;
+        EOS
+        allow(File).to receive(:read).and_return(sql_definition)
+
+        definition = Fx::Definition.new(name: "test", version: 1)
+
+        expect(definition.to_sql.delete("\n").squeeze(' ').strip).to eq erb_sql_definition
```
Comment (teoljungberg) [inline spec/fx/definition_spec.rb:62]: This expectation is more than a mouthful to read. Not a blocker for now, but I'll take a stab at reformatting it to make it leaner and meaner

**Review by teoljungberg — APPROVED** [review]

**Comment by teoljungberg** [comment]
I ran CI @gmhawash - and there are some `standardrb` fixes to be made before we merge this.

**Review by teoljungberg — CHANGES_REQUESTED** [review]
As mentioned above - CI is failing.

**Comment by gmhawash** [comment]
When I run this locally everything passes.  What do I need to do to make it fail locally so I can fix these issues?

**Comment by teoljungberg** [comment]
@gmhawash it’s because your local version of standard.rb is out of date with ci

File: spec/fx/definition_spec.rb [hunk 6]
```diff
@@ -29,13 +29,46 @@
...
+          CREATE OR REPLACE FUNCTION west()
+          RETURNS text AS $$
+          BEGIN
+            RETURN 'west';
+          END;
+          $$ LANGUAGE plpgsql;
+        EOS
+        allow(File).to receive(:read).and_return(sql_definition)
+
+        definition = Fx::Definition.new(name: "test", version: 1)
```
Comment (teoljungberg) [inline spec/fx/definition_spec.rb:60]:
```suggestion
        definition = Fx::Definition.function(name: "test", version: 1)
```

AS this was changed on master with #119

File: spec/fx/definition_spec.rb [hunk 7]
```diff
@@ -29,13 +29,46 @@
...
+          <% %w(test west).each do |function| %>
+            CREATE OR REPLACE FUNCTION <%= function %>()
+            RETURNS text AS $$
+            BEGIN
+              RETURN '<%= function %>';
+            END;
+            $$ LANGUAGE plpgsql;
+          <% end %>
+        EOS
+
```
Comment (teoljungberg) [inline spec/fx/definition_spec.rb:43]:
```suggestion
```

File: lib/fx/definition.rb [hunk 8]
```diff
@@ -24,7 +24,8 @@ def path
     end
 
     def version
-      @version.to_s.rjust(2, "0")
+      @version = latest_version if @version == :latest
```
Comment (teoljungberg) [inline lib/fx/definition.rb:27]: This feels unrealted to this PR?

**Review by teoljungberg — CHANGES_REQUESTED** [review]

**Comment by teoljungberg** [comment]
Having thought about this more from a consistency perspective, this change would make the database migration non-reversible as the evaluated code with ERB can and will probably change.

I would be surprised if that happened to me, and therefore don't want to include that uncertainty.

More reasoning along the same lines from fx's sister project scenic:
https://github.com/scenic-views/scenic/issues/248

**Pull request closed without merging**